### DISCUSSION TITLE
weasyprint Python module causes Unknown Runtime Error in Flask app

### DISCUSSION THREAD

**Discussion by noah-schechter** [vercel/community#3817] [Help]
### Summary

I'm trying to deploy a simple Flask app that uses the weasyprint Python module. The app runs locally, but the weasyprint module is causing problems on Vercel. Specifically:

- The root loads while hosted on Vercel when the lines involving the weasyprint module are commented out. The site, excluding the parts that involve weasyprint, is functional on Vercel when weasyprint isn't imported.
- Adding in the lines `from weasyprint import HTML, CSS` and `from weasyprint.text.fonts import FontConfiguration` cause the site to Error 500 on Vercel. I don't have to execute any other lines involving the weasyprint library to receive the error; the import statements alone are sufficient.
- The logs show only an Unknown Runtime Error. 

Some additional context: weasyprint requires pango, a C module. Vercel's documentation indicates pango is installed on their machines so I don't take any steps to install it in my build process.

I understand this is an incredibly niche problem and may be my mistake, but I'd appreciate any guidance. Thanks for your help!

### Source Code
Here's a simplified replica of the problem, involving three files:
app.py:
```
from flask import Flask

# These lines currently cause the unknown runtime error
from weasyprint import HTML, CSS
from weasyprint.text.fonts import FontConfiguration


app = Flask(__name__)

@app.route('/')
def index():
    return "Hello World"
```

vercel.json:
```
{
    "builds": [
      {
        "src": "app.py",
        "use": "@vercel/python"
      }
    ],
    "routes": [
      {
        "src": "/(.*)",
        "dest": "app.py"
      }
    ]
 }
```

requirements.txt (Note that this is the original requirements.txt file, not the file that reflects the simplified scenario I've included above, so it includes libraries that aren't imported in app.py):
```
annotated-types==0.5.0
anyio==3.7.1
blinker==1.6.2
Brotli==1.0.9
certifi==2023.7.22
cffi==1.15.1
charset-normalizer==3.2.0
click==8.1.7
cssselect2==0.7.0
DateTime==5.2
exceptiongroup==1.1.3
fastapi==0.103.0
Flask==2.3.3
fonttools==4.42.1
html5lib==1.1
idna==3.4
importlib-metadata==6.8.0
itsdangerous==2.1.2
Jinja2==3.1.2
MarkupSafe==2.1.3
Pillow==10.0.0
pycparser==2.21
pydantic==2.3.0
pydantic_core==2.6.3
pydyf==0.7.0
pyphen==0.14.0
pytz==2023.3
requests==2.31.0
six==1.16.0
sniffio==1.3.0
starlette==0.27.0
tinycss2==1.2.1
typing_extensions==4.7.1
urllib3==2.0.4
weasyprint==59.0
webencodings==0.5.1
Werkzeug==2.3.7
zipp==3.16.2
zope.interface==6.0
zopfli==0.2.2
```
If the full source would be helpful, it's available [here](https://github.com/noah-schechter/checklist). Note that the site isn't complete—there are some other steps I need to take to make the PDF writing process work on Vercel, but I'm putting those off until I solve this problem. 

### Error message
```
Unknown application error occurred
Runtime.Unknown
````

**Comment by Waoweens** [comment]
i have the same issue, in my requirements.txt i only have weasyprint, and using the built in BaseHTTPRequestHandler to handle requests

**Comment by Waoweens** [comment]
i figured out how to get full logs (`vercel logs` cli)

<details>
<summary>View logs</summary>

```py
WeasyPrint could not import some external libraries. Please carefully follow the installation steps before reporting an issue:
https://doc.courtbouillon.org/weasyprint/stable/first_steps.html#installation
https://doc.courtbouillon.org/weasyprint/stable/first_steps.html#troubleshooting 
-----
[ERROR] OSError: cannot load library 'gobject-2.0-0': gobject-2.0-0: cannot open shared object file: No such file or directory.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'gobject-2.0-0'
Traceback (most recent call last):
  File "/var/lang/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/var/task/vc__handler__python.py", line 13, in <module>
    __vc_spec.loader.exec_module(__vc_module)
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "./api/pdf.py", line 3, in <module>
    from weasyprint import HTML, CSS
  File "/var/task/weasyprint/__init__.py", line 387, in <module>
    from .css import preprocess_stylesheet  # noqa isort:skip
  File "/var/task/weasyprint/css/__init__.py", line 25, in <module>
    from . import computed_values, counters, media_queries
  File "/var/task/weasyprint/css/computed_values.py", line 11, in <module>
    from ..text.ffi import ffi, pango, units_to_double
  File "/var/task/weasyprint/text/ffi.py", line 428, in <module>
    gobject = _dlopen(
  File "/var/task/weasyprint/text/ffi.py", line 417, in _dlopen
    return ffi.dlopen(names[0])  # pragma: no cover
  File "/var/task/cffi/api.py", line 150, in dlopen
    lib, function_cache = _make_ffi_library(self, name, flags)
  File "/var/task/cffi/api.py", line 832, in _make_ffi_library
    backendlib = _load_backend_lib(backend, libname, flags)
  File "/var/task/cffi/api.py", line 827, in _load_backend_lib
Unknown application error occurred
Runtime.Unknown
```

</details>

looks like it's missing some native libraries, specifically `gobject` but there could be more

**Comment by ti-pmnh** [comment]
> ### Summary
> I'm trying to deploy a simple Flask app that uses the weasyprint Python module. The app runs locally, but the weasyprint module is causing problems on Vercel. Specifically:
> 
> * The root loads while hosted on Vercel when the lines involving the weasyprint module are commented out. The site, excluding the parts that involve weasyprint, is functional on Vercel when weasyprint isn't imported.
> * Adding in the lines `from weasyprint import HTML, CSS` and `from weasyprint.text.fonts import FontConfiguration` cause the site to Error 500 on Vercel. I don't have to execute any other lines involving the weasyprint library to receive the error; the import statements alone are sufficient.
> * The logs show only an Unknown Runtime Error.
> 
> Some additional context: weasyprint requires pango, a C module. Vercel's documentation indicates pango is installed on their machines so I don't take any steps to install it in my build process.
> 
> I understand this is an incredibly niche problem and may be my mistake, but I'd appreciate any guidance. Thanks for your help!
> 
> ### Example
> _No response_
> 
> ### Steps to Reproduce
> ### Source Code
> Here's a simplified replica of the problem, involving three files: app.py:
> 
> ```
> from flask import Flask
> 
> # These lines currently cause the unknown runtime error
> from weasyprint import HTML, CSS
> from weasyprint.text.fonts import FontConfiguration
> 
> 
> app = Flask(__name__)
> 
> @app.route('/')
> def index():
>     return "Hello World"
> ```
> 
> vercel.json:
> 
> ```
> {
>     "builds": [
>       {
>         "src": "app.py",
>         "use": "@vercel/python"
>       }
>     ],
>     "routes": [
>       {
>         "src": "/(.*)",
>         "dest": "app.py"
>       }
>     ]
>  }
> ```
> 
> requirements.txt (Note that this is the original requirements.txt file, not the file that reflects the simplified scenario I've included above, so it includes libraries that aren't imported in app.py):
> 
> ```
> annotated-types==0.5.0
> anyio==3.7.1
> blinker==1.6.2
> Brotli==1.0.9
> certifi==2023.7.22
> cffi==1.15.1
> charset-normalizer==3.2.0
> click==8.1.7
> cssselect2==0.7.0
> DateTime==5.2
> exceptiongroup==1.1.3
> fastapi==0.103.0
> Flask==2.3.3
> fonttools==4.42.1
> html5lib==1.1
> idna==3.4
> importlib-metadata==6.8.0
> itsdangerous==2.1.2
> Jinja2==3.1.2
> MarkupSafe==2.1.3
> Pillow==10.0.0
> pycparser==2.21
> pydantic==2.3.0
> pydantic_core==2.6.3
> pydyf==0.7.0
> pyphen==0.14.0
> pytz==2023.3
> requests==2.31.0
> six==1.16.0
> sniffio==1.3.0
> starlette==0.27.0
> tinycss2==1.2.1
> typing_extensions==4.7.1
> urllib3==2.0.4
> weasyprint==59.0
> webencodings==0.5.1
> Werkzeug==2.3.7
> zipp==3.16.2
> zope.interface==6.0
> zopfli==0.2.2
> ```
> 
> If the full source would be helpful, it's available [here](https://github.com/noah-schechter/checklist). Note that the site isn't complete—there are some other steps I need to take to make the PDF writing process work on Vercel, but I'm putting those off until I solve this problem.
> 
> ### Error message
> ```
> Unknown application error occurred
> Runtime.Unknown
> ```

Hello, were you able to deploy it to Vercel with weasyprint??
I saw that you made an adjustment to build.sh, did it work?
If you can answer I would appreciate it!!!

**Reply by noah-schechter** [reply]
I wasn't able to get this working but I did make some progress. With the help of one of the maintainers of WeasyPrint, I learned:
- WeasyPrint has dependencies outside of Python's package manager, so we need to install them ourselves. 
- Vercel's instructions for installing system libraries are here: https://vercel.com/docs/deployments/build-image#installing-additional-packages. 
- A list of these dependencies are available here: https://github.com/kotify/cloud-print-utils/blob/671001a04bacd78f2dcff9b6c9d04b916a57d086/weasyprint/layer_builder.sh#L8-L17
- Vercel uses Amazon's Linux 2 base image, which uses Pango 1.42, which means that it's only possible to use up to and including WeasyPrint 52 on Vercel. 

The changes I made in my build commands were intended to install these system libraries. I don't remember what the outcome here was; it's definitely a step in the right direction but I eventually put the project on hold.

Let me know if you figure this one out and good luck!
